runtime/vm/ValueTypeHelpers.cpp

```diff
@@ -69,4 +70,106 @@ isClassRefQtype(J9Class *cpContextClass, U_16 cpIndex)
+{
+        /* first field indicates the number of classes in the cache */
+        UDATA length = flattenedClassCache->numberOfEntries;
+        J9Class *clazz = NULL;
+
+        for (UDATA i = 0; i < length; i++) {
+                J9UTF8* currentClassName = J9ROMCLASS_CLASSNAME(J9_VM_FCC_ENTRY_FROM_FCC(flattenedClassCache, i)->clazz->romClass);
+                if (J9UTF8_DATA_EQUALS(J9UTF8_DATA(currentClassName), J9UTF8_LENGTH(currentClassName), className, classNameLength)) {
+                        clazz = J9_VM_FCC_ENTRY_FROM_FCC(flattenedClassCache, i)->clazz;
+                        break;
+                }
+        }
+
+        Assert_VM_notNull(clazz);
+        return clazz;
+}
+
+UDATA
+getFlattenableFieldOffset(J9Class *fieldOwner, J9ROMFieldShape *field)
+{
+        UDATA fieldOffset = 0;
+        J9FlattenedClassCache *flattenedClassCache = NULL;
+        J9ROMNameAndSignature *nameAndSig;
+        UDATA fieldIndex = 0;
+
+        nameAndSig = &field->nameAndSignature;
+        flattenedClassCache = fieldOwner->flattenedClassCache;
+        fieldIndex = findIndexInFlattenedClassCache(flattenedClassCache, nameAndSig);
+        Assert_VM_unequal(UDATA_MAX, fieldIndex);
+        J9FlattenedClassCacheEntry * flattenedClassCacheEntry = J9_VM_FCC_ENTRY_FROM_FCC(flattenedClassCache, fieldIndex);
+        fieldOffset = flattenedClassCacheEntry->offset;
+        return fieldOffset;
+}
+
+BOOLEAN
+isFlattenableFieldFlattened(J9Class *fieldOwner, J9ROMFieldShape *field)
+{
+        BOOLEAN fieldFlattened = FALSE;
+        J9Class* clazz = getFlattenableFieldType(fieldOwner, field);
+        fieldFlattened = J9_ARE_ALL_BITS_SET((clazz)->classFlags, J9ClassIsFlattened);
+        return fieldFlattened;
+}
+
+J9Class *
+getFlattenableFieldType(J9Class *fieldOwner, J9ROMFieldShape *field)
+{
+        J9Class * fieldType = NULL;
+        J9FlattenedClassCache *flattenedClassCache = NULL;
+        J9ROMNameAndSignature *nameAndSig;
+        UDATA fieldIndex = 0;
+
+        nameAndSig = &field->nameAndSignature;
+        flattenedClassCache = fieldOwner->flattenedClassCache;
+        fieldIndex = findIndexInFlattenedClassCache(flattenedClassCache, nameAndSig);
+        Assert_VM_unequal(UDATA_MAX, fieldIndex);
+        fieldType = J9_VM_FCC_ENTRY_FROM_FCC(flattenedClassCache, fieldIndex)->clazz;
+        return fieldType;
+}
+
+UDATA
+getFlattenableFieldSize(J9Class *fieldOwner, J9ROMFieldShape *field)
+{
+        UDATA instanceSize = 0;
+        if(isFlattenableFieldFlattened(fieldOwner, field)){
+                J9Class* clazz = getFlattenableFieldType(fieldOwner, field);
+                instanceSize =  (clazz)->totalInstanceSize - (clazz)->backfillOffset;
+        } else{
+                instanceSize = sizeof(fj9object_t);
+        }
+        return instanceSize;
+}
+
+UDATA
+arrayElementSize(J9ArrayClass* arrayClass)
+{
+        UDATA elementSize = 0;
+        elementSize = arrayClass->flattenedElementSize;
+        return elementSize;
```

Comment:
can simply return `arrayClass->flattenedElementSize;`

Reply:
I added an assert_vm_notNull too. Let me know if I should revert that